project/JavalibIRCleaner.scala

```diff
@@ -0,0 +1,404 @@
+ *  JS types in the Scala library, and ensure that there remains no other
+ *  reference to the Scala library.
+ *
+ *  This ensures that the IR of the javalanglib is truly independent of Scala.
+ *
+ *  The main task is to *completely* erase all references to JS types to
+ *  `j.l.Object`. This includes:
+ *
+ *  - Delete (or do not copy over) .sjsir files that define abstract and native
+ *    JS types.
+ *  - Erase references to JS types in method signatures and `TypeRef`s to
+ *    `java.lang.Object`.
+ *  - Eagerly dereference `LoadJSConstructor` and `LoadJSModule` by "inlining"
+ *    the JS load spec of the mentioned class ref.
+ *
+ *  Afterwards, we check that the IR does not contain any reference to classes
+ *  undef the `scala.*` package, with the unfortunate exception of
+ *  `scala.scalajs.runtime.UndefinedBehaviorError`.
+ */
+object JavalibIRCleaner {
+  def cleanIR(libraryDir: File, javalibDir: File, outputDir: File,
+      logger: Logger): Set[Path] = {
+
+    val errorManager = new ErrorManager(logger)
+
+    val libraryDirPath = libraryDir.toPath()
+    val javalibDirPath = javalibDir.toPath()
+    val outputDirPath = outputDir.toPath()
+
+    val libraryIRFiles = listIRFiles(libraryDirPath)
+    val javalibIRFiles = listIRFiles(javalibDirPath)
+
+    val jsTypes = analyzeJSTypes(libraryIRFiles ++ javalibIRFiles)
+
+    val resultBuilder = Set.newBuilder[Path]
+
+    for (irFile <- javalibIRFiles) {
+      import ClassKind._
+
+      val tree = irFile.tree
+      tree.kind match {
+        case Class | ModuleClass | Interface | HijackedClass =>
+          val cleanedTree = cleanTree(tree, jsTypes, errorManager)
+          val outputPath =
+            outputDirPath.resolve(javalibDirPath.relativize(irFile.path))
+          writeIRFile(outputPath, cleanedTree)
+          resultBuilder += outputPath
+
+        case AbstractJSType | NativeJSClass | NativeJSModuleClass =>
+          // discard
+
+        case JSClass | JSModuleClass =>
+          errorManager.reportError(
+              s"found non-native JS class ${tree.name.encodedName}")(tree.pos)
+      }
+    }
+
+    if (errorManager.hasErrors) {
+      throw new MessageOnlyException(
+          s"There were ${errorManager.errorCount} errors while " +
+          "postprocessing the IR of the javalanglib. " +
+          "The javalanglib must be written in a style that does not leak any " +
+          "reference to the Scala library.")
+    }
+
+    resultBuilder.result()
+  }
+
+  private final class ErrorManager(logger: Logger) {
+    private var _errorCount: Int = 0
+
+    def reportError(msg: String)(implicit pos: Position): Unit = {
+      logger.error(s"$msg at $pos")
+      _errorCount += 1
+    }
+
+    def hasErrors: Boolean = _errorCount != 0
+
+    def errorCount: Int = _errorCount
+  }
+
+  private final class IRFile(val path: Path, val tree: ClassDef)
+
+  private def listIRFiles(path: Path): Seq[IRFile] = {
```

Comment:
We might want to consider exposing this in `IRContainer` (maybe in `standard._` only).

Reply:
See my other comment: I would like to avoid introducing a dependency on `org.scalajs.linker` in the IR cleaner.